### PR TITLE
feat: Workspace Proxy picker show latency to each proxy

### LINKED ISSUE #7381: workspace proxies: show latency indicator on dashboard
Screenshot 2023-05-02 at 11 53 30 AM

It would be nice to see the latency of each proxy and have the Coder dashboard auto-select a proxy for me (given that proxy is healthy)

### PR DESCRIPTION
**Pull request by Emyrk**
**Future work is to auto select based on this latency. That will be a follow up PR**

# What this does

This adds latency checks to workspace proxies in the dashboard. Fixes: https://github.com/coder/coder/issues/7381

(Obviously this looks terrible, but it works!)

Screenshot from 2023-05-11 09-45-52

Cloudflare uses the Performance API in their metrics.

https://blog.cloudflare.com/browser-performance-api/

## New `latency-check` route

This new route does not use any of our middleware. This is intentional to keep the request overhead as minimal as possible. This route is both on the proxy and coderd.

This route also adds a `Timing-Allow-Origin` header which is required to make this performance api in the browser work.

https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/Resource_timing

## Cors to Workspace Proxy

Include the CORs middleware to new `latency-check` route on the workspace proxy.
Add custom headers that we attach to all requests.

### Custom `X-LATENCY-CHECK` header

A `GET` request is a [`simple web request`](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests) which does not trigger CORs preflight on some browsers. Adding a custom header forces it to do a preflight.

## `useProxyLatency.ts` hook to calculate proxy latencies

Uses the performance API to measure latency requests. We measure from `requestStart` to `responseStart`. Which gives us as close to `ping` type measuring as we can get.

timestamp-diagram

These values are only allowed if `Timing-Allow-Origin` is set correctly. If the performance API cannot access these values, it falls back to `duration` which is the total time of the request. I use a [`console.log`](https://github.com/coder/coder/blob/stevenmasley%2Fproxy_picker/site/src/contexts/useProxyLatency.ts#L73) to log this case. We might want to display something in the ui about this.

---

# My development process if anyone was curious.

# Approach 1: Axios timing

My first idea was to time the axios requests to measure latency. This turned out to be flawed. Not only are we measure the overhead of things like TLS, but the axios timing is just imprecise. Even in the interceptors. The values were off my 10ms to the network panel for the total request time.

# Approach 2: Performance API

The better approach is to use: https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/Resource_timing

## Issue 1: Cors + CSP

Setup CORs and CSP to allow external requests to the proxy.

Fixed https://github.com/coder/coder/pull/7484

## Issue 2: The performance api is missing the timing values `requestStart` and `responseStart`

The requestStart value is not coming through. 

Currently trying to use `Timing-Allow-Origin` header, but not getting any results.

https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/requestStart#cross-origin_timing_information

### Notes:

On chrome I was running with `--disable-web-security` to ignore some CORs issues, but that was removing `requestStart` even from local requests which I was using to test the logic. So it looks like that was causing the `requestStart` to always be `0` even if the Timing header was present.

### CODE REVIEW

File: site/src/contexts/useProxyLatency.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  // at is when the latency was recorded.
+  at: Date
+}
+
+interface ProxyLatencyAction {
+  proxyID: string
+  report: ProxyLatencyReport
+}
+
+const proxyLatenciesReducer = (
```
Comment (BrunoQuaresma) [inline site/src/contexts/useProxyLatency.ts:23]: Why did you decide to use a reducer instead of `useState`? Just curious.

**Comment by BrunoQuaresma** [comment]
The stories are showing a `?` in the "latency" column. I think it is because we are not passing the latency value in the mock. Also, could be better to display a híphen instead of a question mark or a helpful message if that is possible.

Screen Shot 2023-05-11 at 10 55 48

File: site/src/contexts/useProxyLatency.ts (hunk 1 above)
Comment (Emyrk) [reply site/src/contexts/useProxyLatency.ts:23]: The state is a `Record<string, ProxyLatencyReport>` for each proxy in our list. Each proxy record is inserted asynchronously that is triggered from a http request.

This async addition to the state made me concerned about any possible bugs from doing this record insertion manually. It feels like a better design pattern for each proxy callback to just `dispatch` their **own latency report.** 

A `useState` would be like:
```typescript
const [state, setState] = useState<Record<string, ProxyLatencyReport>>({})

const observer = new PerformanceObserver((list) => {
  // If we get entries via this callback, then dispatch the events to the latency reducer.
  list.getEntries().forEach((entry) => {
    setState({
    ...state,
    proxy.id: { /* report */}
    })
  })
})
```

So `useState` just means I am pushing this state management to the caller, where the reducer allows the caller to only be concerned about their personal addition to the state, and they do not have to worry about any previous state.

**Comment by BrunoQuaresma** [comment]
Ah, I see. I'm more than happy to help with that, could you please file a ticket for the "improve UI" part and assign it to me please?

**Review by BrunoQuaresma — APPROVED** [review]
FE looks good

**Pull request merged**